File: 0695-max-area-of-island/0695-max-area-of-island.cpp
class Solution {
    // Time Complexity : O(N * M), where N is size of grid's rows and M is size of grid's columns
    // Space Complexity : O(N * M), because we call the recursive function throughout the grid
    // and it use the stack for every recursive call
private:
    int getArea(vector<vector<int>>& grid, int i, int j, int m, int n) {
        if (i < 0 || j < 0 || i >= m || j >= n || grid[i][j] < 1) {
            return 0;
        }
        grid[i][j] = -1; // visited
        int left = getArea(grid, i, j - 1, m, n);
        int right = getArea(grid, i, j + 1, m, n);
        int up = getArea(grid, i - 1, j, m, n);
        int down = getArea(grid, i + 1, j, m, n);
        return left + right + up + down + 1;
    }
    
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        if (grid.size() == 0) return 0;
        int m = grid.size(), n = grid[0].size();
        int maxArea= 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int area = getArea(grid, i, j, m ,n);
                maxArea = max(maxArea, area);
            }
        }
        return maxArea;
    }
};

File: 0695-max-area-of-island/0695-max-area-of-island.java
class Solution {
    private int dfs(int[][] grid, int i, int j, int m, int n) {
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] != 1) {
            return 0;
        }
        // mark as visited
        grid[i][j] = -1;
        int l = dfs(grid, i, j-1, m, n);
        int r = dfs(grid, i, j+1, m, n);
        int u = dfs(grid, i-1, j, m, n);
        int d = dfs(grid, i+1, j, m, n);
        return l + r + u + d + 1;
    }

    public int maxAreaOfIsland(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int maxArea = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    int currentArea = dfs(grid, i, j, m, n);
                    maxArea = Math.max(maxArea, currentArea);
                }
            }
        }
        return maxArea;
    }
}